### PR TITLE
Move IFluidSerializer to shared-object-base

### LINKED ISSUE #8045: Rename IFluidSerializer replaceHandles+decode methods
## Work Item
"replaceHandles" is an ambiguous name and doesn't indicate that it is doing the converse of "decode". They should be encode + decode or encodeHandles + decodeHandles.

**Approach**
Rename one or both to make it clearer what they do and their relationship to each other.

**Open questions**
Should this be gradually deprecated or done all at once in a single version?

**Additional context**
The addition of the decode method is new (it existed on FluidSerializer but not interface).

### PR DESCRIPTION
**Pull request by scarlettjlee**
Copy `IFluidSerializer` to shared-object-base from core-interfaces. (The one in core-interfaces is marked deprecated; it can't be removed yet since it's referenced by deprecated `IFluidObject`.)

Rename `replaceHandles` to `encode` to match `decode` naming.

Make `decode` method required.

Remove deprecated `IFluidDataStoreRuntime.IFluidSerializer`.

Closes #8045. For further context, see comments in #8140.

### CODE REVIEW

File: packages/dds/ink/package.json [hunk 1]
```diff
@@ -60,6 +60,7 @@
     "@fluidframework/datastore-definitions": "^0.55.0",
     "@fluidframework/driver-utils": "^0.55.0",
     "@fluidframework/protocol-definitions": "^0.1026.0",
+    "@fluidframework/runtime-utils": "^0.55.0",
```
Comment (anthony-murphy) [inline packages/dds/ink/package.json:63]: why did we choose runtime-utils for this? I'm wondering if datastore-definitions would be better

Comment (anthony-murphy) [reply packages/dds/ink/package.json:63]: or if its really only needed by dds then shared-object-base

Comment (anthony-murphy) [reply packages/dds/ink/package.json:63]: it seems like even the implementation in runtime-utils can move to share object

File: packages/runtime/runtime-utils/src/serializer.ts [hunk 2]
```diff
@@ -9,12 +9,61 @@
...
+    // Marker to indicate to JSON.parse that the object is a Fluid handle
+    type: "__fluid_handle__";
+
+    // URL to the object. Relative URLs are relative to the handle context passed to the stringify.
+    url: string;
+}
+
+export const IFluidSerializer: keyof IProvideFluidSerializer = "IFluidSerializer";
+
+export interface IProvideFluidSerializer {
```
Comment (scarlettjlee) [inline packages/runtime/runtime-utils/src/serializer.ts:30]: Do we need IProvideFluidSerializer? I suspect we don't.

File: packages/dds/ink/package.json (hunk 1 above)
Comment (scarlettjlee) [reply packages/dds/ink/package.json:63]: I put it in runtime-utils to go with FluidSerializer. shared-object-base makes sense.

**Comment by vladsud** [comment]
> Copy IFluidSerializer to runtime-utils from core-interfaces

I wonder if @fluidframework/shared-object-base should be renamed to be broader and is a better home for it.
For example, I do see serializeHandles() being there in utils.ts (plus SummarySerializer next to it) and that seems like the right place for this interface.

I just worry about runtime-utils to be a dumping ground for all kinds of things, related or not. It seems like "dds utils" is a better bucket to have, and shared-object-base just happened to be it (with not great name, but the right place it seems).

File: packages/dds/merge-tree/src/test/testSerializer.ts [hunk 3]
```diff
@@ -17,13 +15,17 @@ export class TestSerializer implements IFluidSerializer {
...
-    public replaceHandles(
+    public encode(
         input: any,
         bind: IFluidHandle,
     ) {
         throw new Error("Method not implemented.");
     }
 
+    public decode(input: any): any {
+		throw new Error("Method not implemented.");
```
Comment (vladsud) [inline packages/dds/merge-tree/src/test/testSerializer.ts:26]: spacing

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 4]
```diff
@@ -141,10 +139,6 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         return this.dataStoreContext.IFluidHandleContext;
     }
 
-    private readonly serializer: IFluidSerializer;
-    // Back compat: deprecated in 0.53, can be removed in versions >= 0.55.
-    public get IFluidSerializer() { return this.serializer; }
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:146]: You might want to have deprecation in separate PR that goes first in - a bit cleaner.

File: packages/runtime/runtime-utils/src/serializer.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/runtime-utils/src/serializer.ts:30]: I do not see any usage, so the answer should be no - we should delete it.

**Review by vladsud — APPROVED** [review]
Other than location of the code (see comment), looks reasonable to me

**Comment by scarlettjlee** [comment]
I'm going to keep the name shared-object-base for the package. I think dds-utils is too vague and doesn't make more sense to contain SharedObject.

File: common/lib/core-interfaces/src/serializer.ts [hunk 5]
```diff
@@ -6,6 +6,7 @@ import { IFluidHandle } from "./handles";
 
 /**
  * JSON serialized form of an IFluidHandle
+ * @deprecated - Moved to fluidframework/runtime-utils package
```
Comment (anthony-murphy) [inline common/lib/core-interfaces/src/serializer.ts:9]:
```suggestion
 * @deprecated - Moved to fluidframework/shared-object-base package
```

File: packages/dds/shared-object-base/package.json [hunk 6]
```diff
@@ -12,18 +12,24 @@
...
     "build:genver": "gen-version",
+    "build:test": "tsc --project ./src/test/tsconfig.json",
     "ci:build:docs": "api-extractor run --typescript-compiler-folder ../../../node_modules/typescript && copyfiles -u 1 ./_api-extractor-temp/* ../../../_api-extractor-temp/",
     "clean": "rimraf dist lib *.tsbuildinfo *.build.log",
     "eslint": "eslint --format stylish src",
     "eslint:fix": "eslint --format stylish src --fix",
     "lint": "npm run eslint",
     "lint:fix": "npm run eslint:fix",
+    "test": "npm run test:mocha",
+    "test:coverage": "nyc npm test -- --reporter xunit --reporter-option output=nyc/junit-report.xml",
```
Comment (anthony-murphy) [inline packages/dds/shared-object-base/package.json:30]: i think there is a block of nyc stuff that needs to be copied over. you can look at the client build to ensure the test shows up in the test results

**Review by anthony-murphy — APPROVED** [review]

**Review by DLehenbauer — APPROVED** [review]
...aww, merge conflicts.  :-(

But pre-approving because this looks great.

**Pull request merged**